open_spiel/games/gin_rummy.cc

```diff
@@ -0,0 +1,699 @@
+    for (int i = 0; i < kNumPlayers; ++i)
+      deadwood_[i] = TotalCardValue(hands_[i]);
+    knocked_[cur_player_] = true;
+    prev_player_ = cur_player_;
+    phase_ = Phase::kKnock;
+  } else {
+    SPIEL_CHECK_TRUE(std::find(hands_[cur_player_].begin(),
+        hands_[cur_player_].end(), action) != hands_[cur_player_].end());
+    RemoveFromHand(cur_player_, action);
+    deadwood_[cur_player_] = MinDeadwood(hands_[cur_player_]);
+    upcard_ = action;
+    prev_player_ = cur_player_;
+    cur_player_ = Opponent(prev_player_);
+    if (upcard_ == prev_upcard_) {
+      if (repeated_move_) {
+        phase_ = Phase::kGameOver;
+        return;
+      } else {
+        repeated_move_ = true;
+      }
+    } else {
+      repeated_move_ = false;
+    }
+    if (stock_size_ == kWallStockSize) {
+      phase_ = Phase::kWall;
+    } else {
+      phase_ = Phase::kDraw;
+    }
+  }
+}
+
+void GinRummyState::ApplyKnockAction(Action action) {
+  // First the knocking player must discard.
+  if (hands_[cur_player_].size() == 11) {
+    SPIEL_CHECK_TRUE(std::find(hands_[cur_player_].begin(),
+        hands_[cur_player_].end(), action) != hands_[cur_player_].end());
+    RemoveFromHand(cur_player_, action);
+    discard_pile_.push_back(action);
+    deadwood_[cur_player_] = TotalCardValue(hands_[cur_player_]);
+    phase_ = Phase::kKnock;
+  } else if (action == kPassAction) {
+    // Here the pass action indicates knocking player is finished laying the
+    // hand. The player's deadwood is now final, and any cards not layed in a
+    // meld are counted towards the deadwood total.
+    deadwood_[cur_player_] = TotalCardValue(hands_[cur_player_]);
+    // Make sure the knocking player has completed a legal knock.
+    SPIEL_CHECK_LE(deadwood_[cur_player_], knock_card_);
+    // If deadwood equals 0 then the player has gin. The opponent is not
+    // allowed to lay off on gin.
+    if (deadwood_[cur_player_] == 0)
+      finished_layoffs_ = true;
+    cur_player_ = Opponent(prev_player_);
+    phase_ = Phase::kLayoff;
+  } else {
+    // Knocking player must declare or "lay" melds, one action at a time.
+    SPIEL_CHECK_LE(action - kMeldActionBase, kNumMeldActions);
+    SPIEL_CHECK_GE(action - kMeldActionBase, 0);
+    layed_melds_[cur_player_].push_back(action - kMeldActionBase);
+    std::vector<int> meld = int_to_meld.at(action - kMeldActionBase);
+    // Upon laying a meld the cards are removed from the player's hand.
+    for (int card : meld) {
+      RemoveFromHand(cur_player_, card);
+    }
+    deadwood_[cur_player_] = TotalCardValue(hands_[cur_player_]);
+    phase_ = Phase::kKnock;
+  }
+}
+
+void GinRummyState::ApplyLayoffAction(Action action) {
+  if (!finished_layoffs_) {
+    if (action == kPassAction) {
+      finished_layoffs_ = true;
+      phase_ = Phase::kLayoff;
+    } else {
+      SPIEL_CHECK_TRUE(std::find(hands_[cur_player_].begin(),
+          hands_[cur_player_].end(), action) != hands_[cur_player_].end());
```

Comment:
might be worth writing a c_contains as shorthand for this. I keep thinking it should be SPIEL_CHECK_NE, but then realizing this is just missing from c++.

Reply:
Found an absl function that already does it! c_linear_search. Incorporated into new commit.